sdk/storage/Azure.Storage.Common/src/Shared/LazyLoadingReadOnlyStream.cs

```diff
@@ -0,0 +1,240 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Globalization;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+namespace Azure.Storage.Shared
+{
+    internal class LazyLoadingReadOnlyStream<T> : Stream
+    {
+        /// <summary>
+        /// The current position within the blob or file.
+        /// </summary>
+        private long _position;
+
+        /// <summary>
+        /// Last known length of underlying blob or file.
+        /// </summary>
+        private long _length;
+
+        /// <summary>
+        /// The number of bytes in the last download call.
+        /// </summary>
+        private int _lastDownloadBytes;
+
+        /// <summary>
+        /// The number of bytes to download per call.
+        /// </summary>
+        private readonly int _bufferSize;
+
+        /// <summary>
+        /// The stream we will download to.
+        /// </summary>
+        private Stream _stream;
```

Comment:
Any reason to pick a `Stream` instead of an `ArrayPool` buffer?

Reply:
We should also override `Dispose` and clean this up.

Reply:
What would be wrong with just creating a byte array and re-using it over and over?  The only time it would be garbage collected would be when the customer is down with the LazyLoadingReadOnlyStream.

What is the official guidance on implementing a Dispose method?  I find the doc confusing.  Do I implement Dispose() or Dispose(bool)?  https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose

Reply:
If we switch to byte[] from the ArrayPool, I assume we would rent it in the constructor (or an initialization method) return it to the pool in `Dispose()`?

Reply:
We chatted about this offline, but the gist is using `ArrayPool` will result in in less memory pressure if somebody creates a lot of these.  I would either rent the buffer in the .ctor or even lazily wait until right before you first need it.  And you don't need the `Dispose(bool)` overload since this is an internal type.

Reply:
How is the array returned to the pool then?

Reply:
Sorry, I misunderstood your question about `Dispose`.  I thought you were talking about adding your own `Dispose(bool)` per the recommendation linked above.  I wasn't paying attention that this is a `Stream` class and already has them both defined.  You should add:

```C#
protected override void Dispose(bool disposing)
{
    // Return the buffer to the pool if we're called from Dispose or a finalizer
    if (_buffer != null)
    {
        _pool.Return(_buffer, clearArray: true);
        _buffer = null;
    }
}
```

+@pakrym to correct me if I'm wrong, but I think we want to return the buffer regardless of whether we're being called by `Dispose` or a finalizer (i.e., not wrapping the body in `if (disposing) { ... }`).  I also don't think we need to bother making it thread-safe since `Stream`s generally aren't.

Reply:
Fixed in next iteration.